content/docs/rfcs/44/README.md

```diff
@@ -0,0 +1,261 @@
+  HistoryQuery query = 2;
+  HistoryResponse response = 3;
++ SttlementRequest sttlement_request = 19;
++ SttlementResponse settlement_response = 20;
+}
+```
+
+- `SttlementResponse`:
+  - `req_size`: the number of messages matching the `HistoryQuery` criteria
+  - `cost`: the cost of the query in Ether
+  - `request_id`: a random request ID, a 64 char hex string
+  - `history_retrieved`: the number of messages retrieved so far
+- `SttlementRequest`:
+  - `with_payment`: A boolean indicating that the querying node wants to pay for the service
+
+## Flow
+
+- The querying node sends a `HistoryRPC` with its desired `HistoryQuery` criteria and the `sttlement_request` whose `with_payment` is set to `true`.
+- The queried node, checks the `with_payment` field of the `sttlement_request`, and if `true` then it determines the number of messages matching the `HistoryQuery` as `req_size`. 
+It also calculates the `cost` of the query in Ether based on its own internal criteria.
+It populates a `settlement_response` with the `req_size`, `cost`, and a random 32 bytes `request_id`.
+It attaches the `settlement_response` to its `HistoryRPC` and sends it back to the querying node. 
+- The queried node keeps listening to the Waku service contract to get an event about the payment of the `request_id`. 
+Once the payment is received, it proceeds with the normal Store protocol flow with a slight difference. 
+That is, it attaches `settlement_response` to its `HistoryRPC` with all the data fields as specified in the step above. 
+It also increases the `history_retrieved` of the `settlement_response` with the size of the  `messages` in the `HistoryResponse`.
+This is to indicate how many messages are transferred so far and how much is left.
+
+
+# Service Credential Structure 
+<!-- Each service token is associated with a specific value namely `amount` determined by its owner. 
+The `amount` is determined based on the cost of the history query.  -->
+Each service credential is structured as below:
+``` 
+commitment = poseidonHash(amount,  pubkey = poseidonHash(privkey),  blinding)
+```
+`amount` determines the value of the credential in terms of Ether.
+The `privkey` is a random `32` bytes value. 
+The `blinding` is a random `32` bytes blinding factor.
+
+There is also an encryption key `enckey` derived from the `privkey`. 
+<!-- the encryption key derivation to be specified -->
+
+## Service Credential Nullifier 
+
+Each service credential is associated with a nullifier that is computed as follows:
+
+```
+nullifier = poseidonHash(commitment, index || 0, signature)
+```
+where `index` indicates the position of the `commitment` in the Merkle tree,
+and the `signature` is generated over the `commitment` and the `index` as:
+```
+signature = poseidonHash(privkey, commitment, index)
+```
+
+
+# Waku Service Contract
+
+## Registered store nodes
+Nodes offering store protocol who are willing to use the private settlement MUST be registered in the Waku service contract.
```

Comment:
Why? This makes it harder to do zero-cost entry too, e.g. if you want to make some crypto by providing a useful service (key insight behind Swarm's SWAP). If we can relax this assumption to e.g. only use new address + e.g. relayers it'd be useful.

Reply:
We can relax it (it is not a critical design choice), but it may be good later for scoring and reputation, and also making sure some two random users can not transact with the contract (it just makes it harder but not impossible).
On the other side, I support the idea of zero entry.